### PR TITLE
Unit test at pop

### PR DESCRIPTION
**Pull request by hieunguyenmsft**
## Proposed changes

Add unit test and some integration flows for At Pop.

## Type of change

- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [x] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: IdentityCore/src/auth_scheme/MSIDAuthenticationScheme.h [hunk 1]
```diff
@@ -44,7 +44,7 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, readonly) NSDictionary *schemeParameters;
 @property (nonatomic, readonly) MSIDCredentialType credentialType;
 @property (nonatomic, readonly) NSString *tokenType;
-@property (nonatomic, readonly) MSIDAccessToken *accessToken;
+@property (nonatomic, readonly) MSIDAccessToken *blankAccessToken;
```
Comment (rohitnarula7176) [inline IdentityCore/src/auth_scheme/MSIDAuthenticationScheme.h:47]: nit : I think I like the previous naming better. Or maybe we cam come with something like "rawAccessToken"

File: IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.h [hunk 2]
```diff
@@ -32,7 +32,8 @@ NS_ASSUME_NONNULL_BEGIN
 
 @interface MSIDAuthenticationSchemePop : MSIDAuthenticationScheme
 
-- (instancetype)initWithSchemeParameters:(NSDictionary *)schemeParameters;
+@property (nonatomic) NSString *kid;
+@property (nonatomic) NSString *req_cnf;
```
Comment (rohitnarula7176) [inline IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.h:36]: We can keep these properties private in the implementation file itself. If you need to access them in test files , you can use KVC like [scheme valueForKey:@"kid"] . If you want to move them to the headers, we will need to mark them as (readonly) in header file and redeclare it as (readwrite) in implementation file.

File: IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m [hunk 3]
```diff
@@ -64,14 +66,23 @@ - (instancetype)initWithSchemeParameters:(NSDictionary *)schemeParameters
         }
         
         NSString *kidJwk = [_req_cnf msidBase64UrlDecode];
+        if (!kidJwk)
```
Comment (rohitnarula7176) [inline IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m:69]: nit: can we use msidIsStringNilOrBlank:kidJwk here as well for consistency.

File: IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m [hunk 4]
```diff
@@ -64,14 +66,23 @@ - (instancetype)initWithSchemeParameters:(NSDictionary *)schemeParameters
...
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Failed to read req_cnf");
+            return nil;
+        }
         NSData *kidData = [kidJwk dataUsingEncoding:NSUTF8StringEncoding];
-        
+        if (!kidData)
```
Comment (rohitnarula7176) [inline IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m:75]: Can we check !kidData && [kidData length] != 0 just to be safe

File: IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m [hunk 5]
```diff
@@ -64,14 +66,23 @@ - (instancetype)initWithSchemeParameters:(NSDictionary *)schemeParameters
...
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Failed to read req_cnf");
+            return nil;
+        }
         NSError *kidReadingError = nil;
         NSDictionary *kidDict = [NSJSONSerialization JSONObjectWithData:kidData options:0 error:&kidReadingError];
         _kid = [kidDict objectForKey:MSID_KID_CACHE_KEY];
         if (!_kid)
         {
-            MSID_LOG_WITH_CTX(MSIDLogLevelError,nil, @"Failed to generate kid from req_cnf, error: %@", MSID_PII_LOG_MASKABLE(kidReadingError));
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"Failed to generate kid from req_cnf, error: %@", MSID_PII_LOG_MASKABLE(kidReadingError));
```
Comment (rohitnarula7176) [inline IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m:85]: Can we use [kidDict msidObjectForKey:MSID_KID_CACHE_KEY ofClass:[NSString class]] for consistency.

**Comment by kaisong1990** [comment]
@hieunguyenmsft Just for my knowledge, will this fix on unit test require creating keychain access group on mac test host app?

File: IdentityCore/src/auth_scheme/MSIDAuthenticationScheme.h (hunk 1 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/auth_scheme/MSIDAuthenticationScheme.h:47]: Addressed, I will just change back to accessToken.

File: IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.h (hunk 2 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.h:36]: Addressed. I changed it back to private properties. Thank you for the suggestion!

File: IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m (hunk 3 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m:69]: Addressed. Updated validation checks in MSIDAuthenticationSchemePop using [NSString msidIsStringNilOrBlank:] for consistency

File: IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m (hunk 5 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m:85]: Addressed. Updated in next commit.

**Comment by hieunguyenmsft** [comment]
> @hieunguyenmsft Just for my knowledge, will this fix on unit test require creating keychain access group on mac test host app?

Hi Kai, these unit tests should not require a keychain access group on mac test host app. 
Following @rohitnarula7176 suggestion this morning, I updated the code to trick these tests into using the old keychain generator (by re-assign the keychain generator after initializing the device pop manager).

File: IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m (hunk 4 above)
Comment (hieunguyenmsft) [reply IdentityCore/src/auth_scheme/MSIDAuthenticationSchemePop.m:75]: Addressed. Will be in next update

**Pull request merged**